### PR TITLE
Increase full_history_ts_low when flush happens during recovery

### PR DESCRIPTION
**Pull request by jowlyzhang**
This PR adds a missing piece for the UDT in memtable only feature, which is to automatically increase `full_history_ts_low` when flush happens during recovery. 

Test Plan:
Added unit test
make all check

### CODE REVIEW

File: db/db_impl/db_impl_open.cc [hunk 1]
```diff
@@ -1722,6 +1722,23 @@ Status DBImpl::WriteLevel0TableForRecovery(int job_id, ColumnFamilyData* cfd,
...
+    size_t ts_sz = ucmp->timestamp_size();
+    if (ts_sz > 0 && !cfd->ioptions()->persist_user_defined_timestamps) {
+      Slice mem_newest_udt = mem->GetNewestUDT();
+      std::string full_history_ts_low = cfd->GetFullHistoryTsLow();
+      if (full_history_ts_low.empty() ||
+          ucmp->CompareTimestamp(mem_newest_udt, full_history_ts_low) >= 0) {
+        std::string new_full_history_ts_low;
+        GetFullHistoryTsLowFromU64CutoffTs(&mem_newest_udt,
+                                           &new_full_history_ts_low);
+        edit->SetColumnFamily(cfd->GetID());
```
Comment (ltamasi) [inline db/db_impl/db_impl_open.cc:1738]: Very minor but this is probably redundant, as the CF id should already be set in the `VersionEdit`

File: util/udt_util.cc [hunk 2]
```diff
@@ -340,4 +341,13 @@ Status ValidateUserDefinedTimestampsOptions(
   return Status::InvalidArgument(
       "Unsupported user defined timestamps settings change.");
 }
+
+void GetFullHistoryTsLowFromU64CutoffTs(Slice* cutoff_ts,
+                                        std::string* full_history_ts_low) {
+  uint64_t cutoff_udt_ts = 0;
+  bool format_res = GetFixed64(cutoff_ts, &cutoff_udt_ts);
+  assert(format_res);
+  (void)format_res;
```
Comment (ltamasi) [inline util/udt_util.cc:350]: With C++17, an alternative to doing this would be to tag `format_res` with the attribute `[[maybe_unused]]`

**Review by ltamasi — APPROVED** [review]
Thanks for the patch @jowlyzhang - LGTM!

We would definitely want to handle this case correctly; on the other hand though, applications using UDT-in-memtable-only will probably want to set `avoid_flush_during_recovery` to preserve the timestamp history, right?

**Comment by jowlyzhang** [comment]
> Thanks for the patch @jowlyzhang - LGTM!
> 
> We would definitely want to handle this case correctly; on the other hand though, applications using UDT-in-memtable-only will probably want to set `avoid_flush_during_recovery` to preserve the timestamp history, right?

Thank you for calling this out! As a benefit of keeping timestamps in WAL, applications using UDT-in-memtable-only may want to set both `avoid_flush_during_shutdown` and `avoid_flush_during_recovery` to be true to carry around the timestamps during DB reopens. I mentioned this to MyRocks teams' engineers but not in the documentation of this feature. Let me add that to the feature's doc. With that said, you are right that we do want to handle this path correctly!

File: db/db_impl/db_impl_open.cc (hunk 1 above)
Comment (jowlyzhang) [reply db/db_impl/db_impl_open.cc]: Good catch!

**Comment by ltamasi** [comment]
> As a benefit of keeping timestamps in WAL, applications using UDT-in-memtable-only may want to set both `avoid_flush_during_shutdown` and `avoid_flush_during_recovery` to be true to carry around the timestamps during DB reopens.

`avoid_flush_during_shutdown` only comes into play if there are writes that have `disableWAL` set, so MyRocks might not need it but yeah, in general that's another case where we might potentially lose timestamp data.

**Pull request closed without merging**